### PR TITLE
[Metal] Add a way to create a device and queue from raw resources in wgpu-hal

### PR DESCRIPTION
**Pull request by AdrianEddy**
**Checklist**

- [x] Run `cargo clippy`.
- [ ] Run `RUSTFLAGS=--cfg=web_sys_unstable_apis cargo clippy --target wasm32-unknown-unknown` if applicable.
- [x] Add change to CHANGELOG.md. See simple instructions inside file.

**Connections**
Related to #3145 and #2320

**Description**
This PR adds `device_from_raw` and `queue_from_raw` in addition to the existing `texture_from_raw`. This allows to get the `MTLDevice` and/or `MTLCommandQueue` from external application, and use them to process external memory/texture in wgpu.
My use case involves OpenFX plugin for DaVinci Resolve video editor, and Resolve provides MTLTexture and MTLCommandQueue pointers, which I then use in a plugin written in Rust and wgpu.

With these 2 new functions, one can create `wgpu::Device` and `wgpu::Queue` from a raw pointer to `MTLCommandQueue` the following way:
```rust
use foreign_types::ForeignType;
use wgpu_hal::api::Metal;

let mtl_command_queue: *mut metal::MTLCommandQueue = ...;

let (device, queue) = unsafe {
    let mtl_queue = metal::CommandQueue::from_ptr(mtl_command_queue);
    let mtl_device = mtl_queue.device();

    adapter.create_device_from_hal(wgpu_hal::OpenDevice::<Metal> {
        device: <Metal as wgpu_hal::Api>::Device::device_from_raw(mtl_device.to_owned(), wgpu::Features::empty()),
        queue: <Metal as wgpu_hal::Api>::Queue::queue_from_raw(mtl_queue)
    }, &wgpu::DeviceDescriptor {
        label: None,
        features: wgpu::Features::empty(),
        limits: wgpu::Limits::default()
    }, None).unwrap()
};
```

**Testing**
I tested these changes with pointers provided by the DaVinci Resolve, and also with pointers provided from Qt and QML renderer, both cases work as intended.

### CODE REVIEW

File: wgpu-hal/src/metal/device.rs [hunk 1]
```diff
@@ -219,6 +219,13 @@ impl super::Device {
         }
     }
 
+    pub unsafe fn device_from_raw(raw: mtl::Device, features: wgt::Features) -> super::Device {
```
Comment (cwfitzgerald) [inline wgpu-hal/src/metal/device.rs:222]: Are there any safety invariants that need to be upheld by these functions? These kind of functions need their invariant well documented.

Comment (AdrianEddy) [reply wgpu-hal/src/metal/device.rs:222]: I am not sure, I'm not familiar with Metal enough. None of the other functions in there have any safety comments though

**Review by cwfitzgerald — CHANGES_REQUESTED** [review]
Looks good enough for now - we need to go through and audit all of these external interface functions later anyway.

**Comment by AdrianEddy** [comment]
hmm, something went wrong here, I think I messed up the branches, give me a moment

**Comment by AdrianEddy** [comment]
Fixed, sorry about that. This change should only add `device_from_raw` and `queue_from_raw` to the metal backend. Please squash when merging

**Pull request merged**